runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobService.java

```diff
@@ -0,0 +1,309 @@
+        return;
+      }
+
+      // create new invocation
+      JobInvocation invocation = invoker.invoke(preparation, request.getStagingToken());
+      String invocationId = invocation.getId();
+      invocation.start();
+      invocations.put(invocationId, invocation);
+      RunJobResponse response =
+          RunJobResponse.newBuilder().setJobId(invocationId).build();
+      responseObserver.onNext(response);
+      responseObserver.onCompleted();
+    } catch (StatusRuntimeException e) {
+      LOG.warn("Encountered Status Exception", e);
+      responseObserver.onError(e);
+    } catch (Exception e) {
+      LOG.error("Encountered Unexpected Exception", e);
+      responseObserver.onError(Status.INTERNAL.withCause(e).asException());
+    }
+  }
+
+  @Override
+  public void getState(
+      GetJobStateRequest request, StreamObserver<GetJobStateResponse> responseObserver) {
+    try {
+      LOG.trace("{} {}", GetJobStateRequest.class.getSimpleName(), request);
+      String invocationId = request.getJobId();
+      JobInvocation invocation = invocations.get(invocationId);
+
+      JobState.Enum state;
+      synchronized (invocation) {
+        state = invocation.getState();
+      }
+
+      GetJobStateResponse response = GetJobStateResponse.newBuilder().setState(state).build();
+      responseObserver.onNext(response);
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      LOG.error("Encountered Unexpected Exception", e);
+      responseObserver.onError(Status.INTERNAL.withCause(e).asException());
+    }
+  }
+
+  @Override
+  public void cancel(CancelJobRequest request, StreamObserver<CancelJobResponse> responseObserver) {
+    try {
+      LOG.trace("{} {}", CancelJobRequest.class.getSimpleName(), request);
+      String invocationId = request.getJobId();
+      JobInvocation invocation = invocations.get(invocationId);
+
+      JobState.Enum state;
+      synchronized (invocation) {
+        invocation.cancel();
+        state = invocation.getState();
+      }
+
+      CancelJobResponse response = CancelJobResponse.newBuilder().setState(state).build();
+      responseObserver.onNext(response);
+      responseObserver.onCompleted();
+    } catch (Exception e) {
+      LOG.error("Encountered Unexpected Exception", e);
+      responseObserver.onError(Status.INTERNAL.withCause(e).asException());
+    }
+  }
+
+  @Override
+  public void getStateStream(
+      GetJobStateRequest request,
+      StreamObserver<GetJobStateResponse> responseObserver) {
+    try {
+      String invocationId = request.getJobId();
+      JobInvocation invocation = invocations.get(invocationId);
+
+      Function<JobState.Enum, GetJobStateResponse> responseFunction =
+          state -> GetJobStateResponse.newBuilder().setState(state).build();
+      TransformStreamObserver<JobState.Enum, GetJobStateResponse> stateObserver =
+          TransformStreamObserver.create(responseFunction, responseObserver);
+      synchronized (invocation) {
```

Comment:
Ditto. Let invocation handle thread safety.

Reply:
Lets unregister when stream breaks.

Reply:
Could you clarify what you mean by this?

Reply:
The client requesting for state might get disconnected. In that case we should unregister the listener as responseObserver will not work.